alertmanagers static_configs no longer working on HEAD

**What did you do?**
Start Prometheus (587dec9e) with the example config (with the example alertmanager target enabled).

**What did you expect to see?**
Status page (http://localhost:9090/status) should show one configured Alertmanager.

**What did you see instead? Under which circumstances?**
Status page does not show any Alertmanagers.

**Environment**

* System information: Linux 4.14.8-1-ARCH x86_64
* Prometheus version:
```
prometheus, version 2.0.0 (branch: HEAD, revision: 587dec9eb970531cddc7f1803d258e72129b5aa0)
  build user:       christian@wuechoo
  build date:       20171229-22:55:19
  go version:       go1.9.2
```


* Prometheus configuration file:
```yaml
# my global config
global:
  scrape_interval:     15s # Set the scrape interval to every 15 seconds. Default is every 1 minute.
  evaluation_interval: 15s # Evaluate rules every 15 seconds. The default is every 1 minute.
  # scrape_timeout is set to the global default (10s).

# Alertmanager configuration
alerting:
  alertmanagers:
  - static_configs:
    - targets:
       - alertmanager:9093

# Load rules once and periodically evaluate them according to the global 'evaluation_interval'.
rule_files:
  # - "first_rules.yml"
  # - "second_rules.yml"

# A scrape configuration containing exactly one endpoint to scrape:
# Here it's Prometheus itself.
scrape_configs:
  # The job name is added as a label `job=<job_name>` to any timeseries scraped from this config.
  - job_name: 'prometheus'

    # metrics_path defaults to '/metrics'
    # scheme defaults to 'http'.

    static_configs:
      - targets: ['localhost:9090']
```

file_sd_configs seems to yield the same result. This may indicate a general issue with alertmanager target discovery in master. I have not tried other discovery methods though. Maybe I am misunderstand something or I am supposed to use different configs now, but for now, this looks like a bug to me as it unexpectedly changes previous behaviour.

I tried bisect'ing this issue but could not identify a single commit due to build issues (`retrieval/manager.go:95:11: undefined: TargetManager`).

The last working and building commit is a8cce41882e5d94f876d93c1b7c29d37a32f58eb.
The first non-working and building commit is 587dec9eb970531cddc7f1803d258e72129b5aa0

```bash
$ git bisect log
# only skipped commits left to test
# possible first bad commit: [587dec9eb970531cddc7f1803d258e72129b5aa0] rebased and resolved conflicts with the new Discovery GUI page
# possible first bad commit: [60ef2016d508545f0e8b575a954899cae37a96e2] add a cancel func to the scrape pool as it is needed in the scrape loop select block
# possible first bad commit: [80182a5d82d0f29fe8ec483f2ec1ae6d10f9f1de] use poolKey as the pool map key to avoid multi dimensional maps
# possible first bad commit: [1ec76d1950862c1cd5263065616212a97670449d] rearange the contexts variables and logic split the groupsMerge function to set and get other small nits
# possible first bad commit: [6ff1d5c51e3cb0e0df58aa90a473317e25c6d4ac] add the scrape manager config reloader handle errors with invalid scrape config
# possible first bad commit: [b0d4f6ee08632e09d0b0cc8ec0b2a0cbf61b5861] resolved merge confilc in main.go
# possible first bad commit: [f2df712166ef4e16e69d8aace76202619483c0f8] updated README
# possible first bad commit: [aca8f85699211c2453f0121a4015975c605fc2dd] fixed the tests
# possible first bad commit: [fe6c544532360a09c3cbdc6d235d6798b366263d] some renaming and comments fixes. remove some select state that is most likely obsoleete and hoepfully doesn't braje anything :) merge targets will sort by Discoverer name so we can have consistent tests for the maps.
# possible first bad commit: [f5c2c5ff8fc8e21f9bfaa23d0f81c8245ba8632e] brake the start provider func so that can run unit tests against it.
# possible first bad commit: [c5cb0d2910692cfd77b36d7d270677f58f55d3f9] simplify naming and API.
# possible first bad commit: [9c61f0e8a0c1d4f83d06e003af610a7d671287b2] scrape pool doesn't rely on context as Stop() needs to be blocking to prevent Scrape loops trying to write to a closed TSDB storage.
# possible first bad commit: [e405e2f1ea2156f76a7e70f5471b88a590a8e24a] refactored discovery

```

According to the commits, this may be related to pull request #3362 (@krasi-georgiev).

--- a/cmd/prometheus/main.go
+++ b/cmd/prometheus/main.go
@@ -16,6 +16,8 @@ package main
 
 import (
 	"context"
+	"crypto/md5"
+	"encoding/json"
 	"fmt"
 	"net"
 	"net/http"
@@ -295,9 +297,12 @@ func main() {
 		func(cfg *config.Config) error {
 			c := make(map[string]sd_config.ServiceDiscoveryConfig)
 			for _, v := range cfg.AlertingConfig.AlertmanagerConfigs {
-				// AlertmanagerConfigs doesn't hold an unique identifier so we use the config pointer as the identifier.
-				// TODO Krasi - Maybe use the slice index as the reference.
-				c[fmt.Sprintf("%p", v)] = v.ServiceDiscoveryConfig
+				// AlertmanagerConfigs doesn't hold an unique identifier so we use the config hash as the identifier.
+				b, err := json.Marshal(v)
+				if err != nil {
+					return err
+				}
+				c[fmt.Sprintf("%x", md5.Sum(b))] = v.ServiceDiscoveryConfig
 			}
 			return discoveryManagerNotify.ApplyConfig(c)
 		},

--- a/cmd/prometheus/main.go
+++ b/cmd/prometheus/main.go
@@ -296,6 +296,7 @@ func main() {
 			c := make(map[string]sd_config.ServiceDiscoveryConfig)
 			for _, v := range cfg.AlertingConfig.AlertmanagerConfigs {
 				// AlertmanagerConfigs doesn't hold an unique identifier so we use the config pointer as the identifier.
+				// TODO Krasi - Maybe use the slice index as the reference.
 				c[fmt.Sprintf("%p", v)] = v.ServiceDiscoveryConfig
 			}
 			return discoveryManagerNotify.ApplyConfig(c)

--- a/cmd/prometheus/main.go
+++ b/cmd/prometheus/main.go
@@ -44,6 +44,7 @@ import (
 	promlogflag "github.com/prometheus/common/promlog/flag"
 	"github.com/prometheus/prometheus/config"
 	"github.com/prometheus/prometheus/discovery"
+	sd_config "github.com/prometheus/prometheus/discovery/config"
 	"github.com/prometheus/prometheus/notifier"
 	"github.com/prometheus/prometheus/promql"
 	"github.com/prometheus/prometheus/retrieval"
@@ -234,11 +235,12 @@ func main() {
 		ctxWeb, cancelWeb = context.WithCancel(context.Background())
 		ctxRule           = context.Background()
 
-		notifier         = notifier.New(&cfg.notifier, log.With(logger, "component", "notifier"))
-		discoveryManager = discovery.NewManager(log.With(logger, "component", "discovery manager"))
-		scrapeManager    = retrieval.NewScrapeManager(log.With(logger, "component", "scrape manager"), fanoutStorage)
-		queryEngine      = promql.NewEngine(fanoutStorage, &cfg.queryEngine)
-		ruleManager      = rules.NewManager(&rules.ManagerOptions{
+		notifier               = notifier.New(&cfg.notifier, log.With(logger, "component", "notifier"))
+		discoveryManagerScrape = discovery.NewManager(log.With(logger, "component", "discovery manager scrape"))
+		discoveryManagerNotify = discovery.NewManager(log.With(logger, "component", "discovery manager notify"))
+		scrapeManager          = retrieval.NewScrapeManager(log.With(logger, "component", "scrape manager"), fanoutStorage)
+		queryEngine            = promql.NewEngine(fanoutStorage, &cfg.queryEngine)
+		ruleManager            = rules.NewManager(&rules.ManagerOptions{
 			Appendable:  fanoutStorage,
 			QueryFunc:   rules.EngineQueryFunc(queryEngine),
 			NotifyFunc:  sendAlerts(notifier, cfg.web.ExternalURL.String()),
@@ -283,7 +285,21 @@ func main() {
 		remoteStorage.ApplyConfig,
 		webHandler.ApplyConfig,
 		notifier.ApplyConfig,
-		discoveryManager.ApplyConfig,
+		func(cfg *config.Config) error {
+			c := make(map[string]sd_config.ServiceDiscoveryConfig)
+			for _, v := range cfg.ScrapeConfigs {
+				c[v.JobName] = v.ServiceDiscoveryConfig
+			}
+			return discoveryManagerScrape.ApplyConfig(c)
+		},
+		func(cfg *config.Config) error {
+			c := make(map[string]sd_config.ServiceDiscoveryConfig)
+			for _, v := range cfg.AlertingConfig.AlertmanagerConfigs {
+				// AlertmanagerConfigs doesn't hold an unique identifier so we use the config pointer as the identifier.
+				c[fmt.Sprintf("%p", v)] = v.ServiceDiscoveryConfig
+			}
+			return discoveryManagerNotify.ApplyConfig(c)
+		},
 		scrapeManager.ApplyConfig,
 		func(cfg *config.Config) error {
 			// Get all rule files matching the configuration oaths.
@@ -332,23 +348,37 @@ func main() {
 		)
 	}
 	{
-		ctxDiscovery, cancelDiscovery := context.WithCancel(context.Background())
+		ctx, cancel := context.WithCancel(context.Background())
 		g.Add(
 			func() error {
-				err := discoveryManager.Run(ctxDiscovery)
-				level.Info(logger).Log("msg", "Discovery manager stopped")
+				err := discoveryManagerScrape.Run(ctx)
+				level.Info(logger).Log("msg", "Scrape discovery manager stopped")
 				return err
 			},
 			func(err error) {
-				level.Info(logger).Log("msg", "Stopping discovery manager...")
-				cancelDiscovery()
+				level.Info(logger).Log("msg", "Stopping scrape discovery manager...")
+				cancel()
+			},
+		)
+	}
+	{
+		ctx, cancel := context.WithCancel(context.Background())
+		g.Add(
+			func() error {
+				err := discoveryManagerNotify.Run(ctx)
+				level.Info(logger).Log("msg", "Notify discovery manager stopped")
+				return err
+			},
+			func(err error) {
+				level.Info(logger).Log("msg", "Stopping notify discovery manager...")
+				cancel()
 			},
 		)
 	}
 	{
 		g.Add(
 			func() error {
-				err := scrapeManager.Run(discoveryManager.SyncCh())
+				err := scrapeManager.Run(discoveryManagerScrape.SyncCh())
 				level.Info(logger).Log("msg", "Scrape manager stopped")
 				return err
 			},
@@ -493,7 +523,7 @@ func main() {
 		// so keep this interrupt after the ruleManager.Stop().
 		g.Add(
 			func() error {
-				notifier.Run()
+				notifier.Run(discoveryManagerNotify.SyncCh())
 				return nil
 			},
 			func(err error) {

--- a/discovery/manager.go
+++ b/discovery/manager.go
@@ -20,7 +20,6 @@ import (
 	"github.com/go-kit/kit/log"
 	"github.com/go-kit/kit/log/level"
 
-	"github.com/prometheus/prometheus/config"
 	sd_config "github.com/prometheus/prometheus/discovery/config"
 	"github.com/prometheus/prometheus/discovery/targetgroup"
 
@@ -101,13 +100,13 @@ func (m *Manager) SyncCh() <-chan map[string][]*targetgroup.Group {
 }
 
 // ApplyConfig removes all running discovery providers and starts new ones using the provided config.
-func (m *Manager) ApplyConfig(cfg *config.Config) error {
+func (m *Manager) ApplyConfig(cfg map[string]sd_config.ServiceDiscoveryConfig) error {
 	err := make(chan error)
 	m.actionCh <- func(ctx context.Context) {
 		m.cancelDiscoverers()
-		for _, scfg := range cfg.ScrapeConfigs {
-			for provName, prov := range m.providersFromConfig(scfg.ServiceDiscoveryConfig) {
-				m.startProvider(ctx, poolKey{setName: scfg.JobName, provider: provName}, prov)
+		for name, scfg := range cfg {
+			for provName, prov := range m.providersFromConfig(scfg) {
+				m.startProvider(ctx, poolKey{setName: name, provider: provName}, prov)
 			}
 		}
 		close(err)

--- a/discovery/manager_test.go
+++ b/discovery/manager_test.go
@@ -24,6 +24,7 @@ import (
 
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/config"
+	sd_config "github.com/prometheus/prometheus/discovery/config"
 	"github.com/prometheus/prometheus/discovery/targetgroup"
 	"gopkg.in/yaml.v2"
 )
@@ -743,7 +744,11 @@ scrape_configs:
 	discoveryManager := NewManager(nil)
 	go discoveryManager.Run(ctx)
 
-	discoveryManager.ApplyConfig(cfg)
+	c := make(map[string]sd_config.ServiceDiscoveryConfig)
+	for _, v := range cfg.ScrapeConfigs {
+		c[v.JobName] = v.ServiceDiscoveryConfig
+	}
+	discoveryManager.ApplyConfig(c)
 
 	_ = <-discoveryManager.SyncCh()
 	verifyPresence(discoveryManager.targets, poolKey{setName: "prometheus", provider: "static/0"}, "{__address__=\"foo:9090\"}", true)
@@ -758,7 +763,11 @@ scrape_configs:
 	if err := yaml.Unmarshal([]byte(sTwo), cfg); err != nil {
 		t.Fatalf("Unable to load YAML config sOne: %s", err)
 	}
-	discoveryManager.ApplyConfig(cfg)
+	c = make(map[string]sd_config.ServiceDiscoveryConfig)
+	for _, v := range cfg.ScrapeConfigs {
+		c[v.JobName] = v.ServiceDiscoveryConfig
+	}
+	discoveryManager.ApplyConfig(c)
 
 	_ = <-discoveryManager.SyncCh()
 	verifyPresence(discoveryManager.targets, poolKey{setName: "prometheus", provider: "static/0"}, "{__address__=\"foo:9090\"}", true)

--- a/notifier/notifier.go
+++ b/notifier/notifier.go
@@ -16,6 +16,7 @@ package notifier
 import (
 	"bytes"
 	"context"
+	"crypto/md5"
 	"encoding/json"
 	"fmt"
 	"net"
@@ -256,9 +257,12 @@ func (n *Notifier) ApplyConfig(conf *config.Config) error {
 
 		ams.metrics = n.metrics
 
-		// The config pointer is used for the map lookup identifier.
-		// TODO Krasi - Maybe use the slice index as the reference.
-		amSets[fmt.Sprintf("%p", cfg)] = ams
+		// The config hash is used for the map lookup identifier.
+		b, err := json.Marshal(cfg)
+		if err != nil {
+			return err
+		}
+		amSets[fmt.Sprintf("%x", md5.Sum(b))] = ams
 	}
 
 	n.alertmanagers = amSets

--- a/notifier/notifier.go
+++ b/notifier/notifier.go
@@ -113,9 +113,8 @@ type Notifier struct {
 	ctx    context.Context
 	cancel func()
 
-	alertmanagers   []*alertmanagerSet
-	cancelDiscovery func()
-	logger          log.Logger
+	alertmanagers map[string]*alertmanagerSet
+	logger        log.Logger
 }
 
 // Options are the configurable parameters of a Handler.
@@ -247,7 +246,7 @@ func (n *Notifier) ApplyConfig(conf *config.Config) error {
 	n.opts.ExternalLabels = conf.GlobalConfig.ExternalLabels
 	n.opts.RelabelConfigs = conf.AlertingConfig.AlertRelabelConfigs
 
-	amSets := []*alertmanagerSet{}
+	amSets := make(map[string]*alertmanagerSet)
 
 	for _, cfg := range conf.AlertingConfig.AlertmanagerConfigs {
 		ams, err := newAlertmanagerSet(cfg, n.logger)
@@ -257,7 +256,8 @@ func (n *Notifier) ApplyConfig(conf *config.Config) error {
 
 		ams.metrics = n.metrics
 
-		amSets = append(amSets, ams)
+		// The config pointer is used for the map lookup identifier.
+		amSets[fmt.Sprintf("%p", cfg)] = ams
 	}
 
 	n.alertmanagers = amSets
@@ -292,11 +292,14 @@ func (n *Notifier) nextBatch() []*Alert {
 }
 
 // Run dispatches notifications continuously.
-func (n *Notifier) Run() {
+func (n *Notifier) Run(tsets <-chan map[string][]*targetgroup.Group) {
+
 	for {
 		select {
 		case <-n.ctx.Done():
 			return
+		case ts := <-tsets:
+			n.reload(ts)
 		case <-n.more:
 		}
 		alerts := n.nextBatch()
@@ -308,6 +311,20 @@ func (n *Notifier) Run() {
 		if n.queueLen() > 0 {
 			n.setMore()
 		}
+	}
+}
+
+func (n *Notifier) reload(tgs map[string][]*targetgroup.Group) {
+	n.mtx.Lock()
+	defer n.mtx.Unlock()
+
+	for id, tgroup := range tgs {
+		am, ok := n.alertmanagers[id]
+		if !ok {
+			level.Error(n.logger).Log("msg", "couldn't sync alert manager set", "err", fmt.Sprintf("invalid id:%v", id))
+			continue
+		}
+		am.sync(tgroup)
 	}
 }
 
@@ -515,9 +532,9 @@ func newAlertmanagerSet(cfg *config.AlertmanagerConfig, logger log.Logger) (*ale
 	return s, nil
 }
 
-// Sync extracts a deduplicated set of Alertmanager endpoints from a list
+// sync extracts a deduplicated set of Alertmanager endpoints from a list
 // of target groups definitions.
-func (s *alertmanagerSet) Sync(tgs []*targetgroup.Group) {
+func (s *alertmanagerSet) sync(tgs []*targetgroup.Group) {
 	all := []alertmanager{}
 
 	for _, tg := range tgs {

--- a/notifier/notifier.go
+++ b/notifier/notifier.go
@@ -257,6 +257,7 @@ func (n *Notifier) ApplyConfig(conf *config.Config) error {
 		ams.metrics = n.metrics
 
 		// The config pointer is used for the map lookup identifier.
+		// TODO Krasi - Maybe use the slice index as the reference.
 		amSets[fmt.Sprintf("%p", cfg)] = ams
 	}
 

--- a/notifier/notifier_test.go
+++ b/notifier/notifier_test.go
@@ -26,6 +26,7 @@ import (
 	"time"
 
 	old_ctx "golang.org/x/net/context"
+	yaml "gopkg.in/yaml.v2"
 
 	config_util "github.com/prometheus/common/config"
 	"github.com/prometheus/common/model"
@@ -33,6 +34,7 @@ import (
 	"github.com/prometheus/prometheus/discovery/targetgroup"
 	"github.com/prometheus/prometheus/pkg/labels"
 	"github.com/prometheus/prometheus/util/httputil"
+	"github.com/prometheus/prometheus/util/testutil"
 )
 
 func TestPostPath(t *testing.T) {
@@ -447,6 +449,53 @@ func TestLabelSetNotReused(t *testing.T) {
 	if !reflect.DeepEqual(tg, makeInputTargetGroup()) {
 		t.Fatal("Target modified during alertmanager extraction")
 	}
+}
+
+func TestReload(t *testing.T) {
+	var tests = []struct {
+		in  *targetgroup.Group
+		out string
+	}{
+		{
+			in: &targetgroup.Group{
+				Targets: []model.LabelSet{
+					{
+						"__address__": "alertmanager:9093",
+					},
+				},
+			},
+			out: "http://alertmanager:9093/api/v1/alerts",
+		},
+	}
+
+	n := New(&Options{}, nil)
+
+	cfg := &config.Config{}
+	s := `
+alerting:
+  alertmanagers:
+  - static_configs:
+`
+	if err := yaml.Unmarshal([]byte(s), cfg); err != nil {
+		t.Fatalf("Unable to load YAML config: %s", err)
+	}
+
+	if err := n.ApplyConfig(cfg); err != nil {
+		t.Fatalf("Error Applying the config:%v", err)
+	}
+
+	tgs := make(map[string][]*targetgroup.Group)
+	for _, tt := range tests {
+
+		tgs[fmt.Sprintf("%p", cfg.AlertingConfig.AlertmanagerConfigs[0])] = []*targetgroup.Group{
+			tt.in,
+		}
+		n.reload(tgs)
+		res := n.Alertmanagers()[0].String()
+
+		testutil.Equals(t, res, tt.out)
+	}
+
 }
 
 func makeInputTargetGroup() *targetgroup.Group {

--- a/notifier/notifier_test.go
+++ b/notifier/notifier_test.go
@@ -15,6 +15,7 @@ package notifier
 
 import (
 	"context"
+	"crypto/md5"
 	"encoding/json"
 	"fmt"
 	"io/ioutil"
@@ -487,7 +488,11 @@ alerting:
 	tgs := make(map[string][]*targetgroup.Group)
 	for _, tt := range tests {
 
-		tgs[fmt.Sprintf("%p", cfg.AlertingConfig.AlertmanagerConfigs[0])] = []*targetgroup.Group{
+		b, err := json.Marshal(cfg.AlertingConfig.AlertmanagerConfigs[0])
+		if err != nil {
+			t.Fatalf("Error creating config hash:%v", err)
+		}
+		tgs[fmt.Sprintf("%x", md5.Sum(b))] = []*targetgroup.Group{
 			tt.in,
 		}
 		n.reload(tgs)

--- a/notifier/notifier_test.go
+++ b/notifier/notifier_test.go
@@ -173,7 +173,10 @@ func TestHandlerSendAll(t *testing.T) {
 			Password: "testing_password",
 		},
 	}, "auth_alertmanager")
-	h.alertmanagers = append(h.alertmanagers, &alertmanagerSet{
+
+	h.alertmanagers = make(map[string]*alertmanagerSet)
+
+	h.alertmanagers["1"] = &alertmanagerSet{
 		ams: []alertmanager{
 			alertmanagerMock{
 				urlf: func() string { return server1.URL },
@@ -183,9 +186,9 @@ func TestHandlerSendAll(t *testing.T) {
 			Timeout: time.Second,
 		},
 		client: authClient,
-	})
+	}
 
-	h.alertmanagers = append(h.alertmanagers, &alertmanagerSet{
+	h.alertmanagers["2"] = &alertmanagerSet{
 		ams: []alertmanager{
 			alertmanagerMock{
 				urlf: func() string { return server2.URL },
@@ -194,7 +197,7 @@ func TestHandlerSendAll(t *testing.T) {
 		cfg: &config.AlertmanagerConfig{
 			Timeout: time.Second,
 		},
-	})
+	}
 
 	for i := range make([]struct{}, maxBatchSize) {
 		h.queue = append(h.queue, &Alert{
@@ -355,7 +358,10 @@ func TestHandlerQueueing(t *testing.T) {
 	},
 		nil,
 	)
-	h.alertmanagers = append(h.alertmanagers, &alertmanagerSet{
+
+	h.alertmanagers = make(map[string]*alertmanagerSet)
+
+	h.alertmanagers["1"] = &alertmanagerSet{
 		ams: []alertmanager{
 			alertmanagerMock{
 				urlf: func() string { return server.URL },
@@ -364,7 +370,7 @@ func TestHandlerQueueing(t *testing.T) {
 		cfg: &config.AlertmanagerConfig{
 			Timeout: time.Second,
 		},
-	})
+	}
 
 	var alerts []*Alert
 
@@ -374,7 +380,8 @@ func TestHandlerQueueing(t *testing.T) {
 		})
 	}
 
-	go h.Run()
+	c := make(chan map[string][]*targetgroup.Group)
+	go h.Run(c)
 	defer h.Stop()
 
 	h.Send(alerts[:4*maxBatchSize]...)

--- a/retrieval/manager.go
+++ b/retrieval/manager.go
@@ -62,9 +62,7 @@ func (m *ScrapeManager) Run(tsets <-chan map[string][]*targetgroup.Group) error 
 		case f := <-m.actionCh:
 			f()
 		case ts := <-tsets:
-			if err := m.reload(ts); err != nil {
-				level.Error(m.logger).Log("msg", "error reloading the scrape manager", "err", err)
-			}
+			m.reload(ts)
 		case <-m.graceShut:
 			return nil
 		}
@@ -129,11 +127,12 @@ func (m *ScrapeManager) Targets() []*Target {
 	return <-targets
 }
 
-func (m *ScrapeManager) reload(t map[string][]*targetgroup.Group) error {
+func (m *ScrapeManager) reload(t map[string][]*targetgroup.Group) {
 	for tsetName, tgroup := range t {
 		scrapeConfig, ok := m.scrapeConfigs[tsetName]
 		if !ok {
-			return fmt.Errorf("target set '%v' doesn't have valid config", tsetName)
+			level.Error(m.logger).Log("msg", "error reloading target set", "err", fmt.Sprintf("invalid config id:%v", tsetName))
+			continue
 		}
 
 		// Scrape pool doesn't exist so start a new one.
@@ -155,6 +154,4 @@ func (m *ScrapeManager) reload(t map[string][]*targetgroup.Group) error {
 			delete(m.scrapePools, name)
 		}
 	}
-
-	return nil
 }

